lib/mini_repo/api_router.ex

```diff
@@ -74,6 +75,129 @@ defmodule MiniRepo.APIRouter do
+           ),
+         true <- File.exists?(package_path) do
+      send_file(conn, 200, package_path)
+    else
+      _ ->
+        send_resp(conn, 404, "not found")
+    end
+  end
+
+  get "/repos/hexpm_mirror/tarballs/:tarball" do
+    tb =
+      String.split(tarball, "-")
+
+      name = Enum.at(tb, 0)
+      version = 
+      Enum.at(tb, 1)
+      |> String.replace(".tar", "")
+
+    if is_configured_on_demand?() do
+      MiniRepo.Mirror.ServerOnDemand.fetch_tarball_if_not_exist(name, version)
+    end
+
+    with {:ok, data_path} <- get_data_dir(:hexpm_mirror),
+         tarbal_path <-
+           Path.join(
+             Application.app_dir(:mini_repo),
+             "#{data_path}/repos/hexpm_mirror/tarballs/#{tarball}"
+           ),
+         true <- File.exists?(tarbal_path) do
+      send_file(conn, 200, tarbal_path)
+    else
+      _ ->
+        send_resp(conn, 404, "not found")
+    end
+  end
+
+  get "/repos/:repo/packages/:name" do
+    with {:ok, data_path} <- get_data_dir(String.to_atom(repo)),
+         package_path <-
+           Path.join(
+             Application.app_dir(:mini_repo),
+             "#{data_path}/repos/#{repo}/packages/#{name}"
+           )
+          do
+      file = File.exists?(package_path)
+      send_file(conn, 200, package_path)
+    else
+      e ->
+        send_resp(conn, 404, "name: #{name}, repo: #{repo} not found")
+    end
+  end
+
+  get "/repos/:repo/tarballs/:tarball" do
+    with {:ok, data_path} <- get_data_dir(String.to_atom(repo)),
+         tarbal_path <-
+           Path.join(
+             Application.app_dir(:mini_repo),
+             "#{data_path}/repos/#{repo}/tarballs/#{tarball}"
+           ),
+         true <- File.exists?(tarbal_path) do
+      send_file(conn, 200, tarbal_path)
+    else
+      _ ->
+        send_resp(conn, 404, "repo: #{repo}, tarbal: #{tarball} not found")
+    end
+  end
+
+  get "/repos/:repo/names/" do
+    with {:ok, data_path} <- get_data_dir(String.to_atom(repo)),
+         names_path <-
+           Path.join(
+             Application.app_dir(:mini_repo),
+             "#{data_path}/repos/#{repo}/names/"
+           ),
+         true <- File.exists?(names_path) do
+      send_file(conn, 200, names_path)
+    else
+      _ ->
+        send_resp(conn, 404, "not found")
+    end
+  end
+
+  get "/repos/:repo/versions/" do
+    with {:ok, data_path} <- get_data_dir(String.to_atom(repo)),
+         versions_path <-
+           Path.join(
+             Application.app_dir(:mini_repo),
+             "#{data_path}/repos/#{repo}/versions/"
+           ),
+         true <- File.exists?(versions_path) do
+      send_file(conn, 200, versions_path)
+    else
+      _ ->
+        send_resp(conn, 404, "not found")
+    end
+  end
+
+  get "/repos/:repo/docs/:tarball" do
+    with {:ok, data_path} <- get_data_dir(String.to_atom(repo)),
+         docs_path <-
+           Path.join(
+             Application.app_dir(:mini_repo),
```

Comment:
This and similar code above ssumes that the data is being stored locally and not using S3.

Reply:
I personally _very much_ advocate a generic way of getting artifacts for particular artifacts from repos though, but this has to be done a bit more generically, I think with e.g. a behaviour for the retrieval. I would like to be able to reference the API endpoint as the upstream repo instead of the S3 bucket as currently required, though I think this design probably has its own work to be done, out of scope of the feature you're actually trying to get in. Nontheless, this assumes that it's a local mirror which ignores S3 which currently supported in mini_repo and has to at least be considered in the design of this all.